components/rc_log/src/android.rs

```diff
@@ -0,0 +1,209 @@
+//!    stop.
+
+use std::{
+    ffi::CString,
+    os::raw::c_char,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        mpsc::{sync_channel, SyncSender},
+        Arc, Mutex,
+    },
+    thread,
+};
+
+use crate::LogLevel;
+
+/// Type of the log callback provided to us by java/swift. Takes the following
+/// arguments:
+///
+/// - Log level (an i32).
+/// - Tag: a (nullable) nul terminated c string. Caller must not free this
+///   string!
+/// - Message: a (non-nullable) nul terminated c string. Caller must not free
+///   this string!
+///
+/// and returns 0 if we should close the thread, and 1 otherwise. This is done
+/// because attempting to call `disable` from within the log callback will
+/// deadlock.
+pub type LogCallback = unsafe extern "C" fn(i32, *const c_char, *const c_char) -> u8;
+
+// TODO: use serde to send this to the other thread as bincode or something,
+// rather than allocating all these strings for every message.
+struct LogRecord {
+    level: LogLevel,
+    tag: Option<CString>,
+    message: CString,
+}
+
+impl<'a, 'b> From<&'b log::Record<'a>> for LogRecord {
+    // XXX important! Don't log in this function!
+    fn from(r: &'b log::Record<'a>) -> Self {
+        let message = format!("{}", r.args());
+        Self {
+            level: r.level().into(),
+            tag: r
+                .module_path()
+                .and_then(|mp| CString::new(mp.to_owned()).ok()),
+            message: crate::string_to_cstring_lossy(message),
+        }
+    }
+}
+
+enum LogMessage {
+    Stop,
+    Record(LogRecord),
+}
+
+pub struct LogAdapterState {
+    // Thread handle for the BG thread. We can't drop this without problems so weu32
+    // prefix with _ to shut rust up about it being unused.
+    handle: Option<std::thread::JoinHandle<()>>,
+    stopped: Arc<Mutex<bool>>,
+    sender: SyncSender<LogMessage>,
+}
+
+pub struct LogSink {
+    sender: SyncSender<LogMessage>,
+    // Used locally for preventing unnecessary work after the `sender`
+    // is closed. Not shared. Not required for correctness.
+    disabled: AtomicBool,
+}
+
+impl log::Log for LogSink {
+    fn enabled(&self, _metadata: &log::Metadata) -> bool {
+        // Really this could just be Acquire but whatever
+        !self.disabled.load(Ordering::SeqCst)
+    }
+
+    fn flush(&self) {}
+    fn log(&self, record: &log::Record) {
+        // Important: we check stopped before writing, which means
+        // it must be set before
+        if self.disabled.load(Ordering::SeqCst) {
+            // Note: `enabled` is not automatically called.
+            return;
+        }
+        // Either the queue is full, or the receiver is closed.
+        // In either case, we want to stop all logging immediately.
+        if self
+            .sender
+            .try_send(LogMessage::Record(record.into()))
+            .is_err()
+        {
+            self.disabled.store(true, Ordering::SeqCst);
```

Comment:
Not true, it can be re-enabled by an explicit call to disable/enable on the other end, the android code has to make that work by remembering that it's returned 0 in an atomic flag, and calling disable later if it has. It's a bit cludgey, tbh.